### ISSUE TITLE
False positive with `warn-unreachable` and `no-strict-optional`

### ISSUE THREAD

**Issue by Garrett-R** [python/mypy#16787]
**Bug Report**

Mypy warns that a line is unreachable, when it actually is reachable.  I don't think this is a Typeshed issue given [the typing](https://github.com/python/typeshed/blob/26e77cbf67ee0e224808a85b18c08c1dda63aaec/stdlib/re.pyi#L242-L245) looks fine there.

**To Reproduce**

[Mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=no-strict-optional%2Cwarn-unreachable&gist=167f81277dcb018bfc9000f1bf357ed1).  Repeated here for convenience:
```python
import re

if not re.search('a', 'b'):
    print('no match')  # This is indeed reachable!
```

**Expected Behavior**

No Mypy errors

**Actual Behavior**

Says the print statement is unreachable, yet if you execute the code, it'll hit the print line.

**Your Environment**

- Mypy version used: 1.8.0
- Mypy command-line flags:  `--no-strict-optional`, `--warn-unreachable`
- Python version used: 3.12

**Comment by AlexWaygood** [comment]
Duplicate of #16718
